tools/run_tests/filter_pull_request_tests.py

```diff
@@ -0,0 +1,253 @@
+  any begin or end triggers
+  :param file_names: list of changed files generated by _get_changed_files()
+  :param starts_with_triggers: tuple of strings to match with beginning of file names
+  :param ends_with_triggers: tuple of strings to match with end of file names
+  :return: safe to skip tests
+  """
+  # convert lists to tuple to pass into str.startswith() and str.endswith()
+  starts_with_whitelist = tuple(starts_with_whitelist)
+  ends_with_whitelist = tuple(ends_with_whitelist)
+  for file_name in file_names:
+    if starts_with_whitelist and not file_name.startswith(starts_with_whitelist) and \
+       ends_with_whitelist and not file_name.endswith(ends_with_whitelist):
+         return False
+  return True
+
+
+def _remove_irrelevant_tests(tests, tag):
+  """
+  Filters out tests by config or language - will not remove sanitizer tests
+  :param tests: list of all tests generated by run_tests_matrix.py
+  :param tag: string representing language or config to filter - "_(language)_" or "_(config)"
+  :return: list of relevant tests
+  """
+  # todo(mattkwong): find a more reliable way to filter tests - don't use shortname
+  return [test for test in tests if
+          tag not in test.shortname or
+          '_msan' in test.shortname or
+          '_asan' in test.shortname or
+          '_tsan' in test.shortname]
+
+
+def _remove_irrelevant_sanitizer_tests(tests, language_tag=""):
+  """
+  Filters out sanitizer tests - can specify a language to filter - this should be c++ only
+  :param tests: list of all tests generated by run_tests_matrix.py
+  :param language_tag: string specifying a language from which to filter sanitizer tests - "_(language)_"
+  :return: list of relevant tests
+  """
+  if language_tag:
+    return [test for test in tests if not language_tag in test.shortname and
+            not '_asan' in test.shortname and
+            not '_msan' in test.shortname and
+            not '_tsan' in test.shortname]
+  else:
+    return [test for test in tests if
+            '_asan' not in test.shortname and
+            '_msan' not in test.shortname and
+            '_tsan' not in test.shortname]
+
+def filter_tests(tests, base_branch):
+  """
+  Filters out tests that are safe to ignore
+  :param tests: list of all tests generated by run_tests_matrix.py
+  :return: list of relevant tests
+  """
+  print("Finding file differences between %s repo and current branch..." % base_branch)
+  changed_files = _get_changed_files(base_branch)
+  for changed_file in changed_files:
+    print(changed_file)
+
+  # Filter core tests
+  skip_core = _can_skip_tests(changed_files,
+                              starts_with_whitelist=_filter_whitelist(starts_with_whitelist, core_starts_with_triggers),
+                              ends_with_whitelist=ends_with_whitelist)
+  if skip_core:
+    tests = _remove_irrelevant_tests(tests, '_c_')
+
+  # Filter c++ tests
+  skip_cpp = _can_skip_tests(changed_files,
```

Comment:
the skip_X part is very repetitive and at some point, wrong triggers will be used for a wrong language.

```
skip_cpp = _can_skip_tests(  /* one day you'll see this */ core_starts_with_triggers)
```

I'd suggest defining and object/class for each test category (core, cpp, ...) and  then 

```
for  testsuite in  [CPP_TESTS, CORE_TESTS, ....]:
  can_skip = can_skip_tests( ....)
  tests = remove_irrelevant_tests( )

````